src/backend/distributed/progress/multi_progress.c

```diff
@@ -0,0 +1,285 @@
+
+	currentProgressDSMHandle = DSM_HANDLE_INVALID;
+}
+
+
+/*
+ * ProgressMonitorList returns the addresses of monitors of ongoing commands, associated
+ * with the given identifier magic number. The function takes a pass in
+ * pg_stat_get_progress_info output, filters the rows according to the given magic number,
+ * and returns the list of addresses of dynamic shared memory segments. Notice that the
+ * caller detach from the attached segments with a call to DetachFromDSMSegments function.
+ */
+List *
+ProgressMonitorList(uint64 commandTypeMagicNumber, List **attachedDSMSegments)
+{
+	/*
+	 * The expected magic number should reside in the first progress field and the
+	 * actual segment handle in the second but the slot ordering is 1-indexed in the
+	 * tuple table slot and there are 3 other fields before the progress fields in the
+	 * pg_stat_get_progress_info output.
+	 */
+	const int magicNumberIndex = 0 + 1 + 3;
+	const int dsmHandleIndex = 1 + 1 + 3;
+
+	/*
+	 * Currently, Postgres' progress logging mechanism supports only the VACUUM,
+	 * operations. Therefore, we identify ourselves as a VACUUM command but only fill
+	 * a couple of the available fields. Therefore the commands that use Citus' progress
+	 * monitoring API will appear in pg_stat_progress_vacuum output.
+	 */
+	text *commandTypeText = cstring_to_text("VACUUM");
+	Datum commandTypeDatum = PointerGetDatum(commandTypeText);
+	Oid getProgressInfoFunctionOid = InvalidOid;
+	TupleTableSlot *tupleTableSlot = NULL;
+	ReturnSetInfo *progressResultSet = NULL;
+	List *monitorList = NIL;
+
+	getProgressInfoFunctionOid = FunctionOid("pg_catalog",
+											 "pg_stat_get_progress_info",
+											 1);
+
+	progressResultSet = FunctionCallGetTupleStore1(pg_stat_get_progress_info,
+												   getProgressInfoFunctionOid,
+												   commandTypeDatum);
+
+	tupleTableSlot = MakeSingleTupleTableSlot(progressResultSet->setDesc);
+
+	/* iterate over tuples in tuple store, and send them to destination */
+	for (;;)
+	{
+		bool nextTuple = false;
+		bool isNull = false;
+		Datum magicNumberDatum = 0;
+		uint64 magicNumber = 0;
+
+		nextTuple = tuplestore_gettupleslot(progressResultSet->setResult,
+											true,
+											false,
+											tupleTableSlot);
+
+		if (!nextTuple)
+		{
+			break;
+		}
+
+		magicNumberDatum = slot_getattr(tupleTableSlot, magicNumberIndex, &isNull);
+		magicNumber = DatumGetUInt64(magicNumberDatum);
+
+		if (!isNull && magicNumber == commandTypeMagicNumber)
+		{
+			Datum dsmHandleDatum = slot_getattr(tupleTableSlot, dsmHandleIndex, &isNull);
+			dsm_handle dsmHandle = DatumGetUInt64(dsmHandleDatum);
+			dsm_segment *dsmSegment = dsm_find_mapping(dsmHandle);
+
+			if (dsmSegment == NULL)
+			{
+				dsmSegment = dsm_attach(dsmHandle);
+			}
+
+			if (dsmSegment != NULL)
+			{
+				void *monitorAddress = dsm_segment_address(dsmSegment);
+				ProgressMonitorData *monitor =
+					MonitorDataFromDSMSegmentAddress(monitorAddress);
```

Comment:
The code above looks suspiciously similar to `GetCurrentProgressMonitor`, is there a way to combine them?